frame/support/src/traits/tasks.rs

```diff
@@ -0,0 +1,40 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Contains the [`Task`] trait, which defines a general-purpose way for defining and executing
+//! service work, and supporting types.
+
+use sp_runtime::DispatchError;
+use sp_std::slice::Iter;
+
+/// A general-purpose trait which defines a type of service work (i.e., work to performed by an
+/// off-chain worker) including methods for enumerating, validating, indexing, and running
+/// tasks of this type.
+pub trait Task<'a>: Sized {
+	/// Inspects the pallet's state and enumerates tasks of this type.
+	fn enumerate() -> Iter<'a, Self>;
```

Comment:
The lifetime doesn't make any sense here. To what should this lifetime be bound? There is no instance this function is called on.

Reply:
Yeah do you know a way to get this to compile without the lifetime? `Iter` seems to require one, and I assume I don't want to use `'static`

Reply:
ah ok so maybe I should switch to the `Box<dyn Iterator<..`

Reply:
```
type Enumeration: sp_std::iter::Iterator<Item=Self>;

fn enumerate() -> Self::Enumeration;
```

Then we don't need `Box` by default, but people can use `Box` if they like.

Reply:
ahh ok very nice thank you